samcli/commands/deploy/command.py

```diff
@@ -210,12 +202,17 @@ def cli(
     project_type,
     cdk_app,
     cdk_context,
-    iac,
-    project,
 ):
     """
     `sam deploy` command entry point
     """
+    click_ctx = click.get_current_context()
+    project_preprocessor = IacProjectResolver(click_ctx)
```

Comment:
The Resolver part is simplified and it will provide only one method to resolve the project.
For the validator, the package/deploy options have extra validations on images and s3 stacks, so they need ad-hoc methods. Like the decorators, we need to add them to the validation lines. The validation order can also be controlled  for these commands.
https://github.com/aws/aws-sam-cli/pull/3273/files#diff-30b075be7ec33583e72b37a0df8296e2b5c9727c8f8913460ec50452962897c2R162